html/browsers/browsing-the-web/back-forward-cache/service-worker-unregister.https.html

```diff
@@ -0,0 +1,62 @@
+<!doctype html>
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="/common/utils.js"></script>
+<script src="/common/dispatcher/dispatcher.js"></script>
+<script src="resources/helper.sub.js"></script>
+<script src="/service-workers/service-worker/resources/test-helpers.sub.js"></script>
+<script>
+
+promise_test(async t => {
+  // Register a service worker and make this page controlled.
+  const workerUrl =
+      'resources/service-worker.js?pipe=header(Service-Worker-Allowed,../)';
+  const registration =
+      await service_worker_unregister_and_register(t, workerUrl, './');
+  t.add_cleanup(_ => registration.unregister());
+  await wait_for_state(t, registration.installing, 'activated');
+  const controllerChanged = new Promise(
+      resolve => navigator.serviceWorker.oncontrollerchange = resolve);
+  await claim(t, registration.active);
+  await controllerChanged;
+
+  const pageA = new RemoteContext(token());
+  const pageB = new RemoteContext(token());
+
+  const urlA = location.origin + executorPath + pageA.context_id;
+  const urlB = originCrossSite + executorPath + pageB.context_id;
+
+  // Open `urlA`.
+  window.open(urlA, '_blank', 'noopener');
+  await pageA.execute_script(waitForPageShow);
+
+  // Navigate to `urlB`.
+  await pageA.execute_script(
+    (url) => prepareNavigation(() => {
+      location.href = url;
+    }),
+    [urlB]);
+  await pageB.execute_script(waitForPageShow);
+
+  // Unregister the service worker when the controlled `pageA` is in BFCache.
+  // This should not crash.
+  await registration.unregister();
+
+  // Back navigate and check whether the page is restored from BFCache.
+  await pageB.execute_script(
+    () => {
+      prepareNavigation(() => { history.back(); });
+    }
+  );
+  await pageA.execute_script(waitForPageShow);
+  await assert_bfcached(pageA);
+
+  // `pageA` should be still controlled after restored from BFCache.
```

Comment:
Hmm, is this the behavior we want?
https://w3c.github.io/ServiceWorker/#navigator-service-worker-unregister

So the registration would be there until bfcache entry is evicted or page brought back to foreground?

Reply:
@rakina, any thoughts?

Reply:
cc @rubberyuzu @wanderview @fergald as they have a better idea about SW interaction with bfcache

Reply:
Friendly ping. Any thoughts?

Reply:
In my understanding, according to the current https://w3c.github.io/ServiceWorker/#navigator-service-worker-unregister the registration is removed so that subsequent navigation are no longer controlled by the service worker, while the page in BFCache still remains controlled (as "currently [controlled](https://w3c.github.io/ServiceWorker/#dfn-control) [service worker client](https://w3c.github.io/ServiceWorker/#dfn-service-worker-client)'s [active service worker](https://html.spec.whatwg.org/multipage/webappapis.html#concept-environment-active-service-worker)'s [containing service worker registration](https://w3c.github.io/ServiceWorker/#dfn-containing-service-worker-registration) is effective"), but not sure.

Reply:
I don't think it matters here what browsers do here, since bfcache + SW is largely undefined. But it would be, IMO, rather odd to keep the page registered while you can't still communicate with it. So I would expect unregister to evict any bfcached pages (well, that is what UAs can do anyhow, since eviction may happen at any point).

Reply:
Oh, interesting. I think service worker unregistration actually already triggers eviction of the controllees in Chrome's implementation: https://source.chromium.org/chromium/chromium/src/+/main:content/browser/service_worker/service_worker_container_host.cc;l=643;drc=b24429781c133fd2482edbb4d0c8c8305031b082

Not sure if this test is testing a different case, or is somehow not testing what it should be. But I guess eviction makes sense in the case linked above?

Reply:
@hiroshige-g, let's update the test to expect eviction after service worker unregistration. @smaug---- , do you have other concerns for the tests? Maybe we can add a `Client.postMessage()` test too, but I don't think that's blocking the rest of the tests here from landing?

Reply:
> I think service worker unregistration actually already triggers eviction of the controllees in Chrome's implementation:

Probably Comments 7 and 8 of https://bugs.chromium.org/p/chromium/issues/detail?id=1204228 is the explanation for this. Because there are still active controlled pages at the time of `unregister()`, the full eviction code doesn't run.

Anyway I updated the test to expect eviction from BFCache on `unregister()`.